src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs

```diff
@@ -346,5 +346,182 @@ public static void WriteTransactionRollbackError(this SqlDiagnosticListener @thi
+        {
+            return DiagnosticTransactionScope.CreateTransactionRollbackScope(@this, isolationLevel, connection, transaction, transactionName, operationName);
+        }
+    }
+
+    internal ref struct DiagnosticScope //: IDisposable //ref structs cannot implement interfaces but the compiler will use pattern matching
+    {
+        private const int CommandOperation = 1;
+        private const int ConnectionOpenOperation = 2;
+
+        private readonly SqlDiagnosticListener _diagnostics;
+        private readonly int _operation;
+        private readonly string _operationName;
+        private readonly Guid _operationId;
+        private readonly object _context1;
+        private readonly object _context2;
+        private Exception _exception;
+
+        private DiagnosticScope(SqlDiagnosticListener diagnostics, int operation, Guid operationsId, string operationName, object context1, object context2)
+        {
+            _diagnostics = diagnostics;
+            _operation = operation;
+            _operationId = operationsId;
+            _operationName = operationName;
+            _context1 = context1;
+            _context2 = context2;
+            _exception = null;
+        }
+
+        public void Dispose()
+        {
+            switch (_operation)
+            {
+                case CommandOperation:
+                    if (_exception != null)
+                    {
+                        _diagnostics.WriteCommandError(_operationId, (SqlCommand)_context1, (SqlTransaction)_context2, _exception, _operationName);
+                    }
+                    else
+                    {
+                        _diagnostics.WriteCommandAfter(_operationId, (SqlCommand)_context1, (SqlTransaction)_context2, _operationName);
+                    }
+                    break;
+
+                case ConnectionOpenOperation:
+                    if (_exception != null)
+                    {
+                        _diagnostics.WriteConnectionOpenError(_operationId, (SqlConnection)_context1, _exception, _operationName);
+                    }
+                    else
+                    {
+                        _diagnostics.WriteConnectionOpenAfter(_operationId, (SqlConnection)_context1, _operationName);
+                    }
+                    break;
+
+                //case ConnectionCloseOperation:
+                //    if (_exception != null)
+                //    {
+                //        _diagnostics.WriteConnectionCloseError(_operationId, (Guid)_context2, (SqlConnection)_context1, _exception, _operationName);
+                //    }
+                //    else
+                //    {
+                //        _diagnostics.WriteConnectionCloseAfter(_operationId, (Guid)_context2, (SqlConnection)_context1, _operationName);
+                //    }
+                //    break;
```

Comment:
I commented them out because they can't be used but chose not to delete them because it wanted it to be clear that they hadn't been omitted they were purposefully not there. It's probably better to do that with a comment.